Enhancements and cleanup in scipy.integrate (attempt #2)

woodscn opened the pull request:
ENH: Enable full_output option for nquad. Currently returns neval only. Convert scipy.integrate.dblquad and scipy.integrate.tplquad to thin wrappers around scipy.integrate.nquad, maintaining API compatibility. Replaces PR #5892, and handles LANL license requirements by hosting a SciPy fork at losalamos/scipy (the "official LANL version").

woodscn commented:
I'm running this option by the appropriate people around here. Take a look and let me know what you think about it.

Eric89GXL commented:
You mention in this PR description a disclaimer about an "official LANL version", so I was expecting to see it somewhere -- but I don't see "LANL" in the diff (or any of the changed files). Is that the idea, that you no longer need to have a disclaimer here at all?

Eric89GXL commented:
Ahh, neat. Sounds good to me.

woodscn commented:
Any further comments/concerns?

woodscn commented:
Are there any holdups here that I should be addressing?

woodscn commented:
Also resolves #5452.

certik commented:
Hi @pv and @rgommers, would you mind looking at this? It's ready to go in.

woodscn commented:
Guys, seriously. It's been 3 months. What's the holdup here? This is not a complicated PR.

pv commented:
Based on a quick look:
- No test coverage for `full_output`.
- 'message' and 'explain' are processed, but not returned. The for loop on line 760 is unnecessary since `out` is not returned, you can just use the returns from quad directly.
- The docstring should not introduce something as "Preliminary", but describe what's there now. What the option actually does is not explained (and Returns section is not updated).
- Conversion of tplquad and dblquad to nquad may have performance implications from Python overhead, maybe useful to check.
- The full_output probably should not be processed in the inner integrations in nquad, if it is not needed (Python overhead).

woodscn commented:
I added brief test coverage for full_output, and I got rid of the extra full_output processing, and I fixed the docstring. 

As far as the "feel" of the performance change, the two routines are quite similar. Since this change eliminates an otherwise redundant code path in quadpack (originally discussed in PR #2510), it reduces the overall maintenance of an (admittedly largely static) piece of scipy.

Neval actually counts the number of times the integrand function is evaluated, so it actually only increments at the innermost level of nquad. Not really any way to get away from that.

woodscn commented:
Also, do I need to worry about CircleCI? It looks like it's failing because it doesn't have things that numpy needs, which I don't think has anything to do with me.

certik commented:
Sorry, I think we forgot about it. Please keep pinging us until it gets in.

Sent from my mobile phone.
On Jun 3, 2016 9:28 PM, "Nathan Woods" <notifications@github.com> wrote:

> Guys, seriously. It's been 3 months. What's the holdup here? This is not a
> complicated PR.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/5906#issuecomment-223671928>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/AABQWOeerWkmrqvwSRofKrb53rZwhumIks5qIIB7gaJpZM4Hl85c>
> .
>

rgommers commented:
> Also, do I need to worry about CircleCI? It looks like it's failing because it doesn't have things that numpy needs, which I don't think has anything to do with me.

No, don't worry about CircleCI - its setup isn't correct yet. It will fail until we'll get it to build merge commits instead of the PR branch.

rgommers commented on scipy/integrate/tests/test_quadpack.py:
```diff
@@ -245,9 +245,10 @@ def opts_basic(*args):
-
+                    opts=[opts_basic, {}, {}, {}], full_output=True)
+        assert_quad(res[:-1], 1.5267454070738635)
+        assert(res[-1]['neval'] == 388962) 
```
Use ``assert_`` (``assert`` gets stripped with ``python -O``).

This test is probably fragile, number of evaluations may change on other platforms for example. Maybe just check that ``res[-1]['neval']`` exists, or check for ``<= 400000`` to allow some variation.

woodscn commented:
Yeah, I didn't love that test either. I made the recommended changes.

rgommers commented on scipy/integrate/quadpack.py:
```diff
@@ -653,12 +658,12 @@ def nquad(func, ranges, args=None, opts=None):
     >>> integrate.nquad(func, [[0,1], [-1,1], [.13,.8], [-.15,1]],
     ...                 opts=[opts0,{},{},{}])
-    (1.5267454070738633, 2.9437360001402324e-14)
+    (1.5267454070738633, 2.9437360001402324e-14, {'neval': 388962})
```
This change is only correct if you add ``full_output=True`` in the line above.

woodscn replied:
*facepalm*

rgommers commented on scipy/integrate/quadpack.py:
```diff
@@ -750,11 +757,19 @@ def integrate(self, *args, **kwargs):
             f = partial(self.integrate, depth=depth+1)
-
-        value, abserr = quad(f, low, high, args=args, **opt)
+        quad_r = quad(f, low, high, args=args, full_output=True, **opt)
```
``full_output=self.full_output``

rgommers replied:
and a few lines lower, move ``infodict = quad_r[2]`` inside ``if self.full_output:``

rgommers commented:
> The ``full_output`` probably should not be processed in the inner integrations in nquad, if it is not needed (Python overhead).

This comment indicates that @pv expects ``neval`` to be the number of outer integrations (I think). The comment is not addressed anyway, ``full_output`` is used in all dimensions. 

And what is returned now is not comepletely clear, a comment above ``self.out_dict['neval'] += infodict['neval']`` would be useful.

rgommers commented:
> Conversion of tplquad and dblquad to nquad may have performance implications from Python overhead, maybe useful to check.

This seems to be OK, impact is minor:

master:

    In [126]: func = lambda t, x: np.exp(-x*t)/t**4

    In [127]: %timeit nquad(func, [[1, np.inf],[0, np.inf]])
    100 loops, best of 3: 8.45 ms per loop

    In [128]: %timeit dblquad(func, 0, np.inf, lambda x: 1, lambda x: np.inf)
    100 loops, best of 3: 8.27 ms per loop

PR 5906:

    In [2]: func = lambda t, x: np.exp(-x*t)/t**4

    In [3]: %timeit nquad(func, [[1, np.inf],[0, np.inf]])
    100 loops, best of 3: 8.39 ms per loop

    In [4]: %timeit dblquad(func, 0, np.inf, lambda x: 1, lambda x: np.inf)
    100 loops, best of 3: 8.47 ms per loop

And for some very simple functions:

    In [11]: %timeit nquad(lambda x,y: x**2 + y**2, [[0, 10], [-5, 5]])
    1000 loops, best of 3: 295 µs per loop

    In [12]: %timeit nquad(lambda x,y: x**2 + y**2, [[0, 10], [-5, 5]], full_output=True)
    1000 loops, best of 3: 332 µs per loop

    In [13]: %timeit nquad(lambda x,y,z: x**2 + y**2 + z, [[0, 10], [-5, 5], [-1, 1]], full_output=True)
    100 loops, best of 3: 7.27 ms per loop

    In [14]: %timeit nquad(lambda x,y,z: x**2 + y**2 + z, [[0, 10], [-5, 5], [-1, 1]])
    100 loops, best of 3: 6.54 ms per loop

rgommers commented:
This PR looks OK to merge to me now. @pv if you don't agree with interpretation of ``neval`` as number of inner loop evals (https://github.com/scipy/scipy/pull/5906#issuecomment-225365720) please comment. Otherwise I'll hit the green button before 0.18.x branching.

woodscn commented:
I'll go ahead and re-squash the PR, then. 







On Sun, Jun 12, 2016 at 6:39 AM -0600, "Ralf Gommers" <notifications@github.com> wrote:












This PR looks OK to merge to me now. @pv if you don't agree with interpretation of neval as number of inner loop evals (#5906 (comment)) please comment. Otherwise I'll hit the green button before 0.18.x branching.



—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or mute the thread.

rgommers commented:
Allright, gave it 5 more days which should be enough - in it goes. Thanks @woodscn (for the PR and for the patience).

certik commented:
Thanks @rgommers for merging! Time for :beers:.
